### ISSUE TITLE
Fat manifests support

### ISSUE THREAD

**Issue by vpereira** [deitch/docker_registry2#69]
Hi,

We have a patch that we have being using since 2020, that add  support to "fat manifests" as described here https://docs.docker.com/registry/spec/manifest-v2-2/#manifest-list

To request a "fat manifest," we need to provide a specific Accept header - application/vnd.docker.distribution.manifest.list.v2+json
(note the .list part)
AFAIK this gem [does not allow specifying the Accept header](https://github.com/deitch/docker_registry2/blob/7b3577a265535576890c9cfce8564d1994b0b05a/lib/registry/registry.rb#L401)
The monkey patch is to send this different Accept header

```
class DockerRegistry2::Registry
  RETRYS = 3

  def blob(repo, digest)
    JSON.parse doget "/v2/#{repo}/blobs/#{digest}"
  end

  def fat_manifest(repo, tag)
    response = doreq('get', "/v2/#{repo}/manifests/#{tag}", {
      'Accept' => 'application/vnd.docker.distribution.manifest.list.v2+json'
    })
    JSON.parse response
  end

  def v2_manifest_digest(repo, tag)
    response = doreq('head', "/v2/#{repo}/manifests/#{tag}", {
      'Accept' => 'application/vnd.docker.distribution.manifest.v2+json'
    })
    response.headers[:docker_content_digest]
  end

  private

  def doreq(type, url, additional_headers = {})
    retry_count = 0
    begin
      response = RestClient::Request.execute(
        method: type,
        url: @base_uri + url,
        headers: headers(payload: nil).merge(additional_headers),
        open_timeout: @open_timeout,
        read_timeout: @read_timeout
      )
    rescue RestClient::NotFound, RestClient::Exceptions::OpenTimeout
      if retry_count >= RETRYS
        raise
      else
        retry_count += 1
        retry
      end
    rescue SocketError
      raise DockerRegistry2::RegistryUnknownException
    rescue RestClient::NotFound => e
      raise DockerRegistry2::NotFound, e
    rescue RestClient::Unauthorized => e
      header = e.response.headers[:www_authenticate]
      method = header.downcase.split(' ')[0]
      case method
      when 'basic'
        response = do_basic_req(type, url, nil, nil)
      when 'bearer'
        response = do_bearer_req(type, url, header, additional_headers)
      else
        raise DockerRegistry2::RegistryUnknownException
      end
    end
    response
  end

  def do_bearer_req(type, url, header, additional_headers = {})
    token = authenticate_bearer(header)
    retry_count = 0
    begin
      response = RestClient::Request.execute(
        method: type,
        url: @base_uri + url,
        headers: headers(payload: nil, bearer_token: token).merge(additional_headers),
        open_timeout: @open_timeout,
        read_timeout: @read_timeout
      )
    rescue SocketError
      raise DockerRegistry2::RegistryUnknownException
    rescue RestClient::Unauthorized
      raise DockerRegistry2::RegistryAuthenticationException
    rescue RestClient::MethodNotAllowed
      raise DockerRegistry2::InvalidMethod
    rescue RestClient::NotFound, RestClient::Exceptions::OpenTimeout
      if retry_count >= RETRYS
        raise
      else
        retry_count += 1
        retry
      end
    end

    response
  end
end
```

So I'm trying to figure out what would be the best way to make it upstream? Is that something that you are willing to support? How could I transform this patch in a PR?

**Comment by vpereira** [comment]
> "fat manifests"? Is that what they call it? I always called the docker mime-type ones "multi-arch manifest" and the OCI mime-type ones "index". 🤷‍♂️

From the linked official document:

```
The manifest list is the “fat manifest” which points to specific image manifests for one or more platforms.
Its use is optional, and relatively few images will use one of these manifests. 
A client will distinguish a manifest list from an image manifest based on the Content-Type 
returned in the HTTP response.
```

Ok, what do you think passing optional headers to the connect via `opts`, like:

```
opts = { open_timeout: 2, read_timeout: 5, extra_headers: {...} }
reg = DockerRegistry2.connect("https://my.registy.corp.com", opts)
```

**Comment by vpereira** [comment]
closed by #73
